### PR TITLE
[Core][RFC] RPC network error chaos test framework

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Inspired by #19970. This PR adds the ability to inject transient network errors to rpc calls. It can be used to test two things:
1. Idempotency of RPC
2. Retry of RPC

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/rpc/rpc_chaos.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+}  // namespace
+
+RpcFailure get_rpc_failure(const std::string &name) {
+  if (RayConfig::instance().testing_rpc_failure().empty()) {
+    return RpcFailure::None;
+  }
+  return _rpc_failure_manager.GetRpcFailure(name);
+}
+
+void init() { _rpc_failure_manager.Init(); }
```
[the remark this answers is not in the document]
Comment (jjyao) [reply src/ray/rpc/rpc_chaos.cc:97]: This is more like (re)init.

File: src/ray/rpc/rpc_chaos.cc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      for (const auto &item :
+           absl::StrSplit(RayConfig::instance().testing_rpc_failure(), ",")) {
+        std::vector<std::string> parts = absl::StrSplit(item, "=");
+        RAY_CHECK_EQ(parts.size(), 2);
+        failable_methods_.emplace(parts[0], std::atoi(parts[1].c_str()));
+      }
+    }
+  }
+
+  RpcFailure GetRpcFailure(const std::string &name) {
```
[the remark this answers is not in the document]
Comment (jjyao) [reply src/ray/rpc/rpc_chaos.cc:59]: `Get` is borrowed from asio_chaos. What about `GenerateRpcFailure`?

**Review by rynewang — APPROVED** [review]

**Review by dentiny — COMMENTED** [review]
Q: why not use interceptor? one of the design goals is fault injection
https://grpc.io/docs/guides/interceptors/

File: src/ray/rpc/rpc_chaos.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      // 50% chance
+      return RpcFailure::None;
+    }
+  }
+
+ private:
+  absl::Mutex mu_;
+  std::mt19937 gen_;
+  // call name -> # remaining failures
+  std::unordered_map<std::string, uint64_t> failable_methods_ ABSL_GUARDED_BY(&mu_);
```
Comment (dentiny) [inline src/ray/rpc/rpc_chaos.cc:91]: I don't think pointer stability is needed here, should use flat hash map wherever possible

File: src/ray/rpc/rpc_chaos.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+static RpcFailureManager _rpc_failure_manager;
```
Comment (dentiny) [inline src/ray/rpc/rpc_chaos.cc:94]: Three things here:
- `static` has the same semantics as anonymous namespace, which is used to constain visibility into a translation unit, not sure why we need both;
- naming doesn't conform google coding style (should have no _ prefix);
- Non trivially destructible object shouldn't be have global lifecycle, and should use `NoDestructor<>` to wrap.

File: src/ray/rpc/rpc_chaos.cc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+}  // namespace
+
+RpcFailure get_rpc_failure(const std::string &name) {
```
Comment (dentiny) [inline src/ray/rpc/rpc_chaos.cc:98]: google coding style encourages CamelCase

File: src/ray/rpc/rpc_chaos.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      return RpcFailure::None;
+    }
+
+    std::uniform_int_distribution<int> dist(0, 3);
+    int rand = dist(gen_);
+    if (rand == 0) {
+      // 25% chance
+      num_remaining_failures--;
+      return RpcFailure::Request;
+    } else if (rand == 1) {
```
Comment (dentiny) [inline src/ray/rpc/rpc_chaos.cc:77]: clang tidy suggests early return: https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

File: src/ray/rpc/rpc_chaos.cc (hunk 3 above)
Comment (jjyao) [reply src/ray/rpc/rpc_chaos.cc:91]: Yea, I don't need pointer stability. I use the stdlib unordered_map because I don't care about performance here.

Comment (dentiny) [reply src/ray/rpc/rpc_chaos.cc:91]: In another words, flat hash map should be the default

**Pull request merged**